### ISSUE TITLE
host_based_auth_requested() fails on Debian

### ISSUE THREAD

**Issue by xBiggs** [ronf/asyncssh#686]
Debian GNU/Linux 12 (bookworm)

```py
async def sftp_dirs():
  async with asyncssh.connect(socket.gethostbyname(hostname), username=username, password=password, kex_algs="+diffie-hellman-group-exchange-sha1", server_host_key_algs="+ssh-rsa", encryption_algs="+aes128-cbc") as conn:
  async with conn.start_sftp_client as sftp:
    return await sftp.listdir()
  
async def main():
  dirs = await sftp_dirs()
  print(dirs)

if __name__ == '__main__':
  asyncio.run(main())
```
I am directly connected by Ethernet from the Debian client to SSH host. The script fails with `socket.gaierror: [Errno -3] Temporary failure in name resolution`. I have had these name resolution errors happen on Debian before and it usually has something to do with the `/etc/hosts` file, but I don't want to edit it.

I was able to fix the error and connect to the host by changing `host_based_auth_requested()` in `connection.py` from:

```py
async def host_based_auth_requested():
  ...
  if client_host is None:
    sockname = self.get_extra_info('sockname')
    if sockname:
      client_host, _ = await self.loop.getnameinfo(cast(SockAddr, sockname), socket.NI_NUMERICSERV)
    else:
      client_host = ''
  ...
```

To:
```py
async def host_based_auth_requested():
  ...
  if client_host is None:
    sockname = self.get_extra_info('sockname')
    if sockname:
      try:
        client_host, _ = await self.loop.getnameinfo(cast(SockAddr, sockname), socket.NI_NUMERICSERV)
      except socket.gaierror:
        client_host = ''
    else:
      client_host = ''
  ...
```

Sorry for any missing info, typed this on mobile

**Comment by ronf** [comment]
This lookup would probably fail when reverse DNS isn't configured for the IP address associated with the socket.

How are you validating the host without a hostname? Are you putting the host IP address in the client known hosts list?

**Comment by xBiggs** [comment]
> This lookup would probably fail when reverse DNS isn't configured for the IP address associated with the socket.
> 
> How are you validating the host without a hostname? Are you putting the host IP address in the known client hosts list?

Yeah, I think the host is already in the known client hosts list because I connected to it with `ssh` prior to using `asyncssh`

**Comment by ronf** [comment]
The server would actually need to be explicitly configured to enable HostBasedAuthentication, and part of that would involve setting up a global known hosts file, usually in something like /usr/local/etc/ssh/ssh_known_hosts. It looks like there is an option to use a user's ~/.ssh/known_hosts as well, but you'd have to have made an outbound connection from the server machine to the client you are trying to authenticate from for that to be used. Whether the known hosts entry contained a hostname or an IP address would depend on what you entered as the client host when you made that outbound connection from the server, or what was manually added to the global known client hosts.

That said, I think allowing host-based auth to work based on IP address seems reasonable when the server can't resolve a hostname for the client, and I think your change looks good. I'm in the middle of some other work on AsyncSSH right now, but once I have finished that I'll pick this up.

Thanks very much!

**Comment by ronf** [comment]
This change is now available as commit c044221 in the "develop" branch. It turned out to be a little more involved, as there was code on the server side also doing a DNS lookup that needed to be changed.

Also, I decided to set the host to the IP address rather than the empty string when reverse DNS fails. There is still a chance for the host to be empty if there's no sockname available at all (such as when using ProxyCommand), but the application can set the client_host explicitly in that case if needed.

I also made a small tweak to the certificate validation to protect against an empty hostname when doing principal validation. That's definitely an edge case, but it seemed good to be defensive there and the change was simple enough.
